File: CasaAPI/Factories/DeviceDtoFactory.cs
using CasaAPI.Dto;
using CasaAPI.Interfaces;

namespace CasaAPI.Factories
{
    public class DeviceDtoFactory
	{
		public readonly Dictionary<string, Func<dynamic, IDeviceDto>> factory;
		public DeviceDtoFactory()
		{
			factory = new Dictionary<string, Func<dynamic, IDeviceDto>>
			{
				{ "Led", device =>
					{
						LedDto dto = new LedDto();
						dto.baseProperties.id = device.id;
						dto.baseProperties.state = device.state;
						dto.baseProperties.voltage = device.voltage;
						dto.baseProperties.amperes = device.amperes;
						dto.brightness = (int)device.brightness;
						return dto;
					}
				},
				{ "Fan", device =>
					{
						FanDto dto = new FanDto();
						dto.speed = (int)device.speed;
						return dto;
					}
				}
			};
		}
	}
}

File: CasaApp/Utils/DeviceDtoFactory.ts
import { Device, Illuminable, Rotatable } from "@/types/Device";

export function createDeviceDto(device: Device) {
  const dto = [device.type, device.pin, device.state];
  if ("brightness" in device) {
    dto.push((device as Illuminable).brightness);
  }
  if ("speed" in device) {
    dto.push((device as Rotatable).speed);
  }
  return dto;
}
